### PR TITLE
[RLlib] Implement Impala Torch in RLModule / Learner stack

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

This PR introduces the IMPALA torch RL Module.
The next step will be to implement APPO in RL Modules.

Requires https://github.com/ray-project/ray/pull/32825

Shares changes with https://github.com/ray-project/ray/pull/33315

### CODE REVIEW

**Comment by ollie-iterators** [comment]
https://github.com/ray-project/ray/pull/33242/files is about removing all default config objects, so I think since __init__.py is empty, that it should not be included in this PR.

File: rllib/algorithms/impala/torch/impala_torch_policy_rlm.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    global_norm,
+)
+from ray.rllib.utils.typing import TensorType
+
+torch, nn = try_import_torch()
+
+logger = logging.getLogger(__name__)
+
+
+class ImpalaTorchPolicyWithRLModule(
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/impala/torch/impala_torch_policy_rlm.py:24]: We don't need a loss function here thanks to our automatic inference of view requirements.

File: rllib/algorithms/impala/torch/vtrace_torch_v2.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    # Swap B and T axes.
+    res = torch.transpose(rs, 1, 0)
+
+    if drop_last:
+        return res[:-1]
+    return res
+
+
+def vtrace_torch(
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/impala/torch/vtrace_torch_v2.py:62]: Note: This code is a blend of the same counterpart in tf and the old torch vtrace code.

File: rllib/algorithms/impala/tests/test_vtrace_v2.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/impala/tests/test_vtrace_v2.py:1]: I've made it so that v-trace tests for specific frameworks are separated from each other.
That is, the inputs and ground truth might be modified (but should not) by the individual methods to test.
This way, we have clear separation between 1) setting up the same ground truth and inputs for all framework-specific tests the same way and 2) the v trace tests themselves.

File: rllib/algorithms/impala/tests/test_impala_learner.py [hunk 3]
```diff
@@ -33,10 +31,13 @@
...
         np.random.uniform(low=0, high=1, size=(frag_length,))
     ).astype(np.float32),
+    SampleBatch.ACTION_DIST_INPUTS: np.random.normal(
+        0, 1, size=(frag_length, 2)
+    ).astype(np.float32),
 }
 
 
-class TestImpalaTfLearner(unittest.TestCase):
+class TestImpalaTorchLearner(unittest.TestCase):
```
Comment (avnishn) [inline rllib/algorithms/impala/tests/test_impala_learner.py:40]: test impala learner

File: rllib/algorithms/impala/tests/test_impala_learner.py [hunk 4]
```diff
@@ -97,10 +92,7 @@ def test_impala_loss(self):
             learner_group = learner_group_config.build()
             learner_group.set_weights(trainer.get_weights())
             results = learner_group.update(train_batch.as_multi_agent())
-
-            learner_group_loss = results[ALL_MODULES]["total_loss"]
-
-            check(learner_group_loss, policy_loss)
+            print(results)
```
Comment (avnishn) [inline rllib/algorithms/impala/tests/test_impala_learner.py:95]: remove print statement

File: rllib/algorithms/impala/tests/test_vtrace_v2.py [hunk 5]
```diff
@@ -26,15 +28,24 @@ def flatten_batch_and_time_dim(t):
 
 
 class TestVtraceRLModule(unittest.TestCase):
-    def test_vtrace_tf2(self):
-        """Tests V-trace-v2 against ground truth data calculated.
+    """Tests V-trace-v2 against ground truth data calculated.
+
+    Our second implementation of V-Trace (V-trace-v2) is optimized for the least
+    amount of flops. The math in the implementation looks nothing like the math in
```
Comment (avnishn) [inline rllib/algorithms/impala/tests/test_vtrace_v2.py:34]: implementation matches the implementation in the original code base.

Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/impala/tests/test_vtrace_v2.py:34]: Can you write up the full sentence for me, please? Not sure how do merge this snippet with what is already there.

File: rllib/algorithms/impala/torch/impala_torch_learner.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    NUM_ENV_STEPS_TRAINED,
+)
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.typing import ResultDict, TensorType
+
+torch, nn = try_import_torch()
+
+
+@dataclass
+class ImpalaHPs(LearnerHPs):
```
Comment (avnishn) [inline rllib/algorithms/impala/torch/impala_torch_learner.py:27]: can we make this shared between the tf and torch classes.

File: rllib/algorithms/impala/torch/impala_torch_learner.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    ) -> Mapping[str, Any]:
+        results = super().compile_results(
+            batch, fwd_out, postprocessed_loss, postprocessed_gradients
+        )
+        results[ALL_MODULES][NUM_AGENT_STEPS_TRAINED] = batch.agent_steps()
+        results[ALL_MODULES][NUM_ENV_STEPS_TRAINED] = batch.env_steps()
+        return results
+
+
+def _reduce_impala_results(results: List[ResultDict]) -> ResultDict:
```
Comment (avnishn) [inline rllib/algorithms/impala/torch/impala_torch_learner.py:201]: can this shared between the tf and torch classes.

File: rllib/algorithms/impala/tests/test_vtrace_v2.py (hunk 5 above)
Comment (avnishn) [reply rllib/algorithms/impala/tests/test_vtrace_v2.py:34]: The current implementation of Vtrace in RLlib matches the one written in deepmind's scalable-agent repo.
It is an implementation that is optimized for reducing the number of FLOPs per vtrace calculation via dynamic programming.

Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/impala/tests/test_vtrace_v2.py]: @avnishn I have taken your suggestion and also made a couple more changes that I think improve the docstrings of our V-Trace classes. Please have a look at the latest commit!

Comment (avnishn) [reply rllib/algorithms/impala/tests/test_vtrace_v2.py]: gr8

File: rllib/algorithms/impala/torch/impala_torch_policy_rlm.py (hunk 1 above)
Comment (avnishn) [reply rllib/algorithms/impala/torch/impala_torch_policy_rlm.py:24]: algorithm implementors don't need one, but in order to merge this, I think it would be better to implement one now, use it to test the learner loss, and advertise it since in 2.4 we won't be advertising the learners.

**Review by avnishn — COMMENTED** [review]
I still think this pr is gonna need the test of the policy with rlm against the learner to be merged.

I don't love it as a testing strategy, so we should brainstorm and come up with something for after 2.4 in terms of a better testing strategy.

**Comment by ArturNiederfahrenhorst** [comment]
No need to review - accidentally pushed a revert of [https://github.com/ray-project/ray/pull/33311](https://github.com/ray-project/ray/pull/33311#issuecomment-1475029953), which is currently troubling me.

File: rllib/algorithms/impala/tf/impala_tf_learner.py [hunk 8]
```diff
@@ -1,93 +1,29 @@
...
-)
 from ray.rllib.utils.framework import try_import_tf
-from ray.rllib.utils.typing import ResultDict, TensorType
+from ray.rllib.utils.typing import TensorType
 
 _, tf, _ = try_import_tf()
 
 
-@dataclass
-class ImpalaHPs(LearnerHPs):
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/impala/tf/impala_tf_learner.py:23]: Moved to base learner file

File: rllib/algorithms/impala/tf/impala_tf_learner.py [hunk 9]
```diff
@@ -143,16 +80,19 @@ def compute_loss_per_module(
             discounts=discounts_time_major,
         )
 
-        batch_size = tf.cast(target_actions_logp_time_major.shape[-1], tf.float32)
+        # Sample size is T x B, where T is the trajectory length and B is the batch size
+        sample_size = tf.cast(
+            tf.math.reduce_prod(target_actions_logp_time_major.shape[:1]), tf.float32
+        )
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/impala/tf/impala_tf_learner.py:86]: @avnishn @kouroshHakha I argue that calculating the mean value per TS is more meaningful here.

File: rllib/algorithms/impala/tf/impala_tf_learner.py [hunk 10]
```diff
@@ -166,40 +106,3 @@ def compute_loss_per_module(
             "pi_loss": mean_pi_loss,
             "vf_loss": mean_vf_loss,
         }
-
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/impala/tf/impala_tf_learner.py:169]: Also moved to base learner.

File: rllib/algorithms/impala/tf/impala_tf_policy_rlm.py [hunk 11]
```diff
@@ -121,12 +121,12 @@ def loss(
...
         # The policy gradients loss.
         pi_loss = -tf.reduce_sum(target_actions_logp_time_major * pg_advantages)
-        mean_pi_loss = -tf.reduce_mean(target_actions_logp_time_major * pg_advantages)
+        mean_pi_loss = pi_loss
 
         # The baseline loss.
         delta = values_time_major - vtrace_adjusted_target_values
         vf_loss = 0.5 * tf.reduce_sum(tf.math.pow(delta, 2.0))
-        mean_vf_loss = 0.5 * tf.reduce_mean(tf.math.pow(delta, 2.0))
+        mean_vf_loss = vf_loss
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/impala/tf/impala_tf_policy_rlm.py:129]: @avnishn Why are we doing this? I know we'll remove this eventually. I'm still curious if this is an oversight or if the means are deliberately not calculated.

File: rllib/algorithms/impala/torch/impala_torch_policy_rlm.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/impala/torch/impala_torch_policy_rlm.py:30]: Done

File: rllib/algorithms/impala/tf/impala_tf_learner.py (hunk 9 above)
Comment (kouroshHakha) [reply rllib/algorithms/impala/tf/impala_tf_learner.py:86]: As we chatted offline, let's keep the regression minimal (close to none) compared to the old stack. add a todo about what needs to be done here, for later.

File: rllib/algorithms/impala/tf/impala_tf_policy_rlm.py [hunk 12]
```diff
@@ -121,12 +121,12 @@ def loss(
 
         # The policy gradients loss.
         pi_loss = -tf.reduce_sum(target_actions_logp_time_major * pg_advantages)
-        mean_pi_loss = -tf.reduce_mean(target_actions_logp_time_major * pg_advantages)
+        mean_pi_loss = pi_loss
```
Comment (kouroshHakha) [inline rllib/algorithms/impala/tf/impala_tf_policy_rlm.py:124]: again pi_loss is sum mean_pi_loss is the mean. Let's not change that.

File: rllib/algorithms/impala/tf/impala_tf_policy_rlm.py (hunk 11 above)
Comment (kouroshHakha) [reply rllib/algorithms/impala/tf/impala_tf_policy_rlm.py:129]: This is deliberate as far as I can tell, to minimize the old and new differences.

File: rllib/algorithms/impala/torch/impala_torch_learner.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        ImpalaBaseLearner.__init__(self, *args, **kwargs)
+
+    @override(TorchLearner)
+    def compute_loss_per_module(
+        self, module_id: str, batch: SampleBatch, fwd_out: Mapping[str, TensorType]
+    ) -> TensorType:
+        target_policy_dist = fwd_out[SampleBatch.ACTION_DIST]
+        values = fwd_out[SampleBatch.VF_PREDS]
+
+        # TODO(Artur): Why are we missing batch[SampleBatch.ACTION_LOGP] here?
```
Comment (kouroshHakha) [inline rllib/algorithms/impala/torch/impala_torch_learner.py:34]: What does this question imply? is this note still relevant?

File: rllib/algorithms/impala/torch/impala_torch_learner.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+            discounts=discounts_time_major,
+            rewards=rewards_time_major,
+            values=values_time_major,
+            bootstrap_value=bootstrap_value,
+            clip_rho_threshold=self.vtrace_clip_rho_threshold,
+            clip_pg_rho_threshold=self.vtrace_clip_pg_rho_threshold,
+        )
+
+        # Sample size is T x B, where T is the trajectory length and B is the batch size
+        sample_size = convert_to_torch_tensor(
```
Comment (kouroshHakha) [inline rllib/algorithms/impala/torch/impala_torch_learner.py:97]: let's make sure this is the same as the old torch impala code. If there are differences even on the old code, let's leave TODOs here.

File: rllib/algorithms/impala/torch/impala_torch_policy_rlm.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+    @Deprecated(new="ImpalaTorchLearner.compute_loss_per_module()", error=False)
+    @override(TorchPolicyV2)
+    def loss(
+        self,
+        model: PPOTorchRLModule,
+        dist_class,
+        train_batch: SampleBatch,
+    ) -> Union[TensorType, List[TensorType]]:
+        train_batch[SampleBatch.ACTIONS]
```
Comment (kouroshHakha) [inline rllib/algorithms/impala/torch/impala_torch_policy_rlm.py:52]: can you leave a comment on why you need to access these keys but not do anything with them?

**Review by kouroshHakha — APPROVED** [review]
Looks good. Can you attach a screen shot of learning behavior comparing old and new stack? Thanks.

File: rllib/algorithms/impala/torch/impala_torch_learner.py (hunk 13 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/impala/torch/impala_torch_learner.py:34]: Sorry, this TODO got resolved.

File: rllib/algorithms/impala/torch/impala_torch_policy_rlm.py (hunk 15 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/impala/torch/impala_torch_policy_rlm.py:52]: As discussed offline, I mirrored the TF implementation here but was now able to remove all of these without breaking the learner test.

**Comment by ArturNiederfahrenhorst** [comment]
Overall throughput over time.
Screenshot 2023-03-23 at 10 00 36

Mean rewards over time, just to prove that we are learning.
This PR also contains tests for the loss functions and equality of V-Trace.
Screenshot 2023-03-23 at 10 49 27

Throughput in connectors.
Screenshot 2023-03-23 at 10 54 09
